### ISSUE TITLE
Wrong result for cast expr when value exceeds range

### ISSUE THREAD

**Issue by guo-shaoge** [pingcap/tidb#32213]
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

    drop table if exists test.t1;
    create table test.t1(c1 float, c2 double);
    insert into test.t1 values(999.99, 999.99);
    select cast(test.t1.c1 as decimal(4, 1)) from test.t1;

### 2. What did you expect to see? (Required)

    mysql> select cast(test.t1.c1 as decimal(4, 1)) from test.t1;
    +-----------------------------------+
    | cast(test.t1.c1 as decimal(4, 1)) |
    +-----------------------------------+
    |                             999.9 |
    +-----------------------------------+
    1 row in set, 1 warning (0.01 sec)
    
    mysql> show warnings;
    +---------+------+----------------------------------------------------------------------------+
    | Level   | Code | Message                                                                    |
    +---------+------+----------------------------------------------------------------------------+
    | Warning | 1264 | Out of range value for column 'cast(test.t1.c1 as decimal(4, 1))' at row 1 |
    +---------+------+----------------------------------------------------------------------------+
    1 row in set (0.00 sec)
### 3. What did you see instead (Required)

    mysql> select cast(test.t1.c1 as decimal(4, 1)) from test.t1;
    +-----------------------------------+
    | cast(test.t1.c1 as decimal(4, 1)) |
    +-----------------------------------+
    |                            1000.0 |
    +-----------------------------------+
    1 row in set, 1 warning (0.00 sec)
### 4. What is your TiDB version? (Required)

    mysql> select tidb_version();
    +----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
    | tidb_version()
                              |
    +----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
    | Release Version: v5.5.0-alpha-146-g7fd298d916
    Edition: Community
    Git Commit Hash: 7fd298d916c30ed0060db3581be58b448325819e
    Git Branch: master
    UTC Build Time: 2022-02-08 06:57:26
    GoVersion: go1.16.4
    Race Enabled: false
    TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
    Check Table Before Drop: false |
    +----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
    1 row in set (0.00 sec)

**Comment by guo-shaoge** [comment]
Also exists for decimal to decimal:

    drop table if exists test.t1;
    create table test.t1(c1 decimal(6, 4));
    insert into test.t1 values(99.9999);
    select cast(test.t1.c1 as decimal(5, 3)) from test.t1;


tidb:

    mysql> select cast(test.t1.c1 as decimal(5, 3)) from test.t1;
    +-----------------------------------+
    | cast(test.t1.c1 as decimal(5, 3)) |
    +-----------------------------------+
    |                           100.000 |
    +-----------------------------------+
    1 row in set, 1 warning (0.00 sec)
    
    mysql> select version();
    +------------------------------------------+
    | version()                                |
    +------------------------------------------+
    | 5.7.25-TiDB-v5.5.0-alpha-146-g7fd298d916 |
    +------------------------------------------+
    1 row in set (0.00 sec)

mysql:

    mysql> select cast(test.t1.c1 as decimal(5, 3)) from test.t1;
    +-----------------------------------+
    | cast(test.t1.c1 as decimal(5, 3)) |
    +-----------------------------------+
    |                            99.999 |
    +-----------------------------------+
    1 row in set, 1 warning (0.01 sec)
    
    mysql> select version();
    +-----------+
    | version() |
    +-----------+
    | 8.0.23    |
    +-----------+
    1 row in set (0.00 sec)

**Comment by gengliqi** [comment]
This bug has existed for a long time. The oldest version I have tested is v3.0.(Maybe this bug exists on all versions)
So change the severity to major.

**Comment by gengliqi** [comment]
I wasted some time comparing the `round` function implementation(a little bit complex) between tidb and MySQL.
I think the bug isn't here.

Finally, I found the bug after comparing the code between https://github.com/mysql/mysql-server/blob/0cd98bdf981583a1cf4cb526581fc16e23bb839b/sql/item_func.cc#L1725-L1751 and https://github.com/pingcap/tidb/blob/e130e522904106ecf9815fdb3aaccad68dcb5ee0/types/datum.go#L1426-L1459.

The root cause is that the judgment of digit length should be after calling `round` function, not before because the decimal may have a longer digit length after calling `round`.
